google/cloud/spanner/integration_tests/data_types_integration_test.cc

```diff
@@ -0,0 +1,101 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/client.h"
+#include "google/cloud/spanner/database.h"
+#include "google/cloud/spanner/database_admin_client.h"
+#include "google/cloud/spanner/mutations.h"
+#include "google/cloud/spanner/testing/database_environment.h"
+#include "google/cloud/internal/getenv.h"
+#include "google/cloud/testing_util/assert_ok.h"
+#include "google/cloud/testing_util/init_google_mock.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace {
+
+class DataTypeIntegrationTest : public ::testing::Test {
+ public:
+  static void SetUpTestSuite() {
+    client_ = google::cloud::internal::make_unique<Client>(
+        MakeConnection(spanner_testing::DatabaseEnvironment::GetDatabase()));
+  }
+
+  void SetUp() override {
+    auto commit_result = client_->Commit([](Transaction const&) {
+      return Mutations{
+          MakeDeleteMutation("Singers", KeySet::All()),
+          MakeDeleteMutation("DataTypes", KeySet::All()),
+      };
+    });
+    EXPECT_STATUS_OK(commit_result);
+  }
+
+  static void TearDownTestSuite() { client_ = nullptr; }
+
+ protected:
+  static std::unique_ptr<Client> client_;
+};
+
+std::unique_ptr<Client> DataTypeIntegrationTest::client_;
+
+TEST_F(DataTypeIntegrationTest, ReadWriteDate) {
+  auto& client = *client_;
+  using RowType = std::tuple<std::string, Date, std::string>;
+  RowType read_back;
+  auto commit_result = client_->Commit(
+      [&client, &read_back](Transaction const& txn) -> StatusOr<Mutations> {
+        auto result = client.ExecuteDml(
+            txn,
+            SqlStatement(
+                "INSERT INTO DataTypes (Id, DateValue, StringValue) "
+                "VALUES(@id, @date, @event)",
+                {{"id", Value("ReadWriteDate-1")},
+                 {"date", Value(Date(161, 3, 8))},
```

Comment:
Perhaps we should add a "> 9999" case to see how they handle their "YYYY: Four-digit year" edict.

Reply:
Anything over 9999 results in an error.  That is consistent with the [documentation](https://cloud.google.com/spanner/docs/data-types#date-type); which says  the valid range is `0001-01-01` to `9999-12-31`